src/Features/CSharp/Portable/ConvertForEachToFor/CSharpConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -25,25 +26,9 @@ public CSharpConvertForEachToForCodeRefactoringProvider()
 
         protected override string Title => CSharpFeaturesResources.Convert_to_for;
 
-        protected override ForEachStatementSyntax GetForEachStatement(TextSpan selection, SyntaxToken token)
-        {
-            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
-            // https://github.com/dotnet/roslyn/issues/30584: Add tests for this scenario
-            if (foreachStatement == null || foreachStatement.AwaitKeyword != default)
-            {
-                return null;
-            }
-
-            // support refactoring only if caret is in between "foreach" and ")"
-            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
-            if (!scope.IntersectsWith(selection))
-            {
-                return null;
-            }
-
-            return foreachStatement;
-        }
-
+        // https://github.com/dotnet/roslyn/issues/30584: Add tests for this scenario
+        protected override bool IsValid(SyntaxNode foreachStatement) 
+            => foreachStatement is ForEachStatementSyntax foreachStatementTyped && foreachStatementTyped.AwaitKeyword != default;
```

Comment:
consider makng abstract type generic and using `<ForEachStatementSyntax>` so you don't need this cast.